src/ram/transform/ChoiceConversion.cpp

```diff
@@ -105,6 +120,37 @@ Own<Operation> ChoiceConversionTransformer::rewriteIndexScan(const IndexScan* in
         return mk<IndexChoice>(rel, identifier, souffle::clone(filter->getCondition()), std::move(newValues),
                 souffle::clone(filter->getOperation()), indexScan->getProfileText());
     }
+
+    // Check that Relation is not referenced further down in the loop nest
+    bool referencedBelow = false;
+    visitDepthFirst(*indexScan, [&](const TupleElement& element) {
+        if (element.getTupleId() == indexScan->getTupleId()) {
+            referencedBelow = true;
+        }
+    });
+
+    // Convert the Scan into a Choice where True
+    if (!referencedBelow) {
+        RamPattern newValues;
+        for (auto& cur : indexScan->getRangePattern().first) {
+            Expression* val = nullptr;
+            if (cur != nullptr) {
+                val = cur->clone();
+            }
+            newValues.first.emplace_back(val);
+        }
+        for (auto& cur : indexScan->getRangePattern().second) {
+            Expression* val = nullptr;
+            if (cur != nullptr) {
+                val = cur->clone();
+            }
+            newValues.second.emplace_back(val);
+        }
+
```

Comment:
A potential minor nitpick:  Admittedly, I am not super familiar with this part of the codebase but from a design perspective, is there a correspondence between the values of `getRangePattern().first` and `.second`?  Is there a one-to-one correspondence between the `nullptrs` in the first sequence and in the second? I may be worth iterating them together and verify the invariant holds? 

Aside for the slight code duplication here, I think this is basically what one of the overloads of `souffle::clone` does.  It may be worth removing the loops completely and replace them with 

```suggestion
souffle::append(newValues.first, souffle::clone(indexScan->getRangePattern().first);
```

or, in this case, since the vector will be empty to start with, just

```suggestion
newValues.first = souffle::clone(indexScan->getRangePattern().first);
```

(and similiarly for `.second`)

Food for thought! :)

Reply:
We were lazy in the type implementation for the patterns. They define vectors for values for lower and upper bounds for attributes in a relation. 

For example, 
```
.decl A(x:number, y:number) 
``` 
defines a relation `A` which can be constrained by a pattern 
``
l1 <= x <= u1 /\ l2 <= y <= u2
```
The first vector represents the values `(l1,l2)` and the second vector of the pattern represents the values `(u1,u2)`. 

I hope this makes sense.